### ISSUE TITLE
Implement TZIP-16 according to TZIP-12

### ISSUE THREAD

**Issue by gromakovsky** [tqtezos/stablecoin#131]
# Clarification and motivation

See https://gitlab.com/tzip/tzip/-/merge_requests/94. Now TZIP-12 (FA2) refers to TZIP-16 and has new requirements. We should update `stablecoin` contract accordingly.

Note: this issue sort of overlaps with #124 because TZIP-17 now also requires TZIP-16 compliance.

# Acceptance criteria

After https://gitlab.com/tzip/tzip/-/merge_requests/94 is merged, `stablcoin` contract should comply with the newest version of FA2.

**Comment by gromakovsky** [comment]
If we should make some updates in FA2 types, it might be better to do #115 before starting this issue, so that we update FA2 types in only one place (in `morley-ledgers`).

**Comment by sras** [comment]
@gromakovsky It appears that we have to implement only the following changes for TZIP-12 compliance.

1. Token_metadata_registery entrypoint and storage field should be removed, and token metadata should be stored in TZIP-16 metadata itself.

2. Implment off chain views for `get_balance`, `total_supply`, `all_tokens`, `is_operator`.

**Comment by gromakovsky** [comment]
> Token_metadata_registery entrypoint and storage field should be removed

Yes.

> token metadata should be stored in TZIP-16 metadata itself.

Apparently there are two options: store it in a `%token_metadata`-annotated `big_map` or as a `token_metadata` off-chain view. The latter should make the storage type smaller, decreasing gas consumption and making operations a bit cheaper. So I think it's preferred. This off-chain view doesn't have to access storage as long as metadata is constant, it can return constant data.

> Impelment off chain views for get_balance, total_supply, all_tokens, is_operator.

These are optional, but I think it's better to provide them anyway. Also if we expose token metadata as an off-chain view, all of these off-chain views will be mandatory (and we'll have to implement `token_metadata` as well).

To sum up, I think we have to make only one smart contract change: remove `token_metadata_registry` and the storage field it uses. On the client side we should extend the metadata storage to contain 5 off-chain views. Besides, we can do some cleanup, e. g. remove the old `metadata.tz` contract.

**Comment by sras** [comment]
>These are optional 

It appears that the spec says

> If present, all of these SHOULD be implemented, at least, as *“Michelson Storage Views”*

Doesn't it mean it should be all or none? So I thought as we are probably implementing `token_metadata` as an off chain view, we should implement others as well.

**Comment by gromakovsky** [comment]
> Doesn't it mean it should be all or none?

It does.

> So I thought as we are probably implementing token_metadata as an off chain view, we should implement others as well.

Yes, that's what I wrote after saying they are optional:
> Also if we expose token metadata as an off-chain view, all of these off-chain views will be mandatory (and we'll have to implement token_metadata as well).

We can have `token_metadata` big_map instead, then these will be optional, but I think implementing 5 off-chain views is a better option.
